test/test_utils/__init__.py

```diff
@@ -1094,21 +1114,61 @@ def delete_uploaded_tests_from_s3(s3_test_location):
 
 
 def get_dlc_images():
-    if is_pr_context() or is_empty_build_context():
+    if is_deep_canary_context():
+        deep_canary_images = get_deep_canary_images(
+            canary_framework=os.getenv("FRAMEWORK"),
+            canary_image_type=get_image_type(),
+            canary_arch_type=get_test_job_arch_type(),
+            canary_region=os.getenv("AWS_REGION"),
+        )
+        return " ".join(deep_canary_images)
+    elif is_pr_context() or is_empty_build_context():
         return os.getenv("DLC_IMAGES")
     elif is_canary_context():
         # TODO: Remove 'training' default once training-specific canaries are added
         image_type = get_image_type() or "training"
         return parse_canary_images(os.getenv("FRAMEWORK"), os.getenv("AWS_REGION"), image_type)
-    test_env_file = os.path.join(
-        os.getenv("CODEBUILD_SRC_DIR_DLC_IMAGES_JSON"), "test_type_images.json"
-    )
-    with open(test_env_file) as test_env:
-        test_images = json.load(test_env)
-    for dlc_test_type, images in test_images.items():
-        if dlc_test_type == "sanity":
-            return " ".join(images)
-    raise RuntimeError(f"Cannot find any images for in {test_images}")
+    elif is_mainline_context():
+        test_env_file = os.path.join(
+            os.getenv("CODEBUILD_SRC_DIR_DLC_IMAGES_JSON"), "test_type_images.json"
+        )
+        with open(test_env_file) as test_env:
+            test_images = json.load(test_env)
+        for dlc_test_type, images in test_images.items():
+            if dlc_test_type == "sanity":
+                return " ".join(images)
+        raise RuntimeError(f"Cannot find any images for in {test_images}")
+    return None
+
+
+def get_deep_canary_images(canary_framework, canary_image_type, canary_arch_type, canary_region):
+    all_images = get_canary_image_uris_from_bucket()
+    matching_images = []
+    for image_uri in all_images:
+        image_framework = get_framework_from_image_uri(image_uri)
+        image_type = get_image_type_from_tag(image_uri)
+        image_arch_type = get_image_arch_type_from_tag(image_uri)
+        image_region = get_region_from_image_uri(image_uri)
+        if (
```

Comment:
Is this enough to consider an image as "matching"? How will this scale to other image flavors?

Reply:
As of now, the scope of deep canaries is limited to `x86` and `arm`, but I see your point. I'll look into how easy/difficult it is to make future-proof support for flavors.